### ISSUE TITLE
Winteger-overflow does not detect overflow on unsigned integers,

### ISSUE THREAD

**Issue by PiotrZSL** [llvm/llvm-project#64828]
Example:
```
std::size_t k1Pb =  1024U * 1024U * 1024U * 1024U * 1024U;
//k1Pb:
//        .quad   0                               # 0x0
```

overflow to 0, not detected by both GCC and Clang, when:
```
std::size_t k1Pb =  1LU * 1024U * 1024U * 1024U * 1024U * 1024U;
//k1Pb:
//        .quad   1125899906842624                # 0x4000000000000
```
is calculated correctly, at same time code like this:
```
std::size_t k1Pb =  1024 * 1024 * 1024 * 1024 * 1024;
```
Trigger -Winteger-overflow in Clang and -Woverflow in GCC.

**Comment by firewave** [comment]
The non-`-Winteger-overflow` cases are detected by the `bugprone-implicit-widening-of-multiplication-result` clang-tidy check:
```
<source>:6:26: warning: performing an implicit widening conversion to type 'std::size_t' (aka 'unsigned long') of a multiplication performed in type 'unsigned int' [bugprone-implicit-widening-of-multiplication-result]
    6 |     std::size_t k1Pb1 =  1024U * 1024U * 1024U * 1024U * 1024U;
      |                          ^
<source>:6:26: note: make conversion explicit to silence this warning
    1 | #include <cstdlib>
    2 | 
    3 | extern void f();
    4 | extern void f()
    5 | {
    6 |     std::size_t k1Pb1 =  1024U * 1024U * 1024U * 1024U * 1024U;
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                          static_cast<std::size_t>(       )
<source>:6:26: note: perform multiplication in a wider type
    6 |     std::size_t k1Pb1 =  1024U * 1024U * 1024U * 1024U * 1024U;
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                          static_cast<std::size_t>( )
<source>:10:26: warning: performing an implicit widening conversion to type 'std::size_t' (aka 'unsigned long') of a multiplication performed in type 'int' [bugprone-implicit-widening-of-multiplication-result]
   10 |     std::size_t k1Pb3 =  1024 * 1024 * 1024 * 1024 * 1024;
      |                          ^
<source>:10:26: note: make conversion explicit to silence this warning
   10 |     std::size_t k1Pb3 =  1024 * 1024 * 1024 * 1024 * 1024;
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                          static_cast<std::size_t>(   )
<source>:10:26: note: perform multiplication in a wider type
   10 |     std::size_t k1Pb3 =  1024 * 1024 * 1024 * 1024 * 1024;
      |                          ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                          static_cast<long>(   )
```
https://godbolt.org/z/739dG9rbG

**Comment by PiotrZSL** [comment]
Yes I know, but there is also: https://github.com/llvm/llvm-project/issues/64732
For me those issues should be detected by a compiler itself.
This is a reason why I'm pointing this out.

**Comment by shafik** [comment]
I have to say this bug feels funny, but maybe I am missing something not obvious, the check for overflow in `CheckedIntArithmetic(...)` is being short-circuited here:

https://github.com/llvm/llvm-project/blob/7ff7df1c629a78f6c87a0bfbc7e5d7c2714283fe/clang/lib/AST/ExprConstant.cpp#L2790-L2793

Basically if the LHS is unsigned don't check for overflow and just return the result of the operation, so this will get diagnosed:

```cpp
std::size_t k1Pb1 =  1024 * 1024 * 1024 * 1024 * 1024U;
```

and funny enough gcc seems to follow the same logic: https://godbolt.org/z/bGW1xrEM6

This looks like a bug to me CC @AaronBallman @tbaederr

**Comment by felix642** [comment]
I can be wrong, but I'm guessing that since unsigned overflow is defined behavior as per the standard (and signed is not) we warn for signed overflow only?

**Comment by shafik** [comment]
> I can be wrong, but I'm guessing that since unsigned overflow is defined behavior as per the standard (and signed is not) we warn for signed overflow only?

Sure, then if should be either the LHS or the RHS then. Perhaps it is assuming the LHS and the RHS are the operands after the usual arithmetic conversions and thus checking either one would be sufficient. 

Since these are apparently not after UAC then checking either for being unsigned would not be sufficient since UAC does not guarantee the result will be unsigned if one of the operands are. 

We would have to check if `Result` is unsigned, at least that is what it looks like to me.

**Comment by AaronBallman** [comment]
> I can be wrong, but I'm guessing that since unsigned overflow is defined behavior as per the standard (and signed is not) we warn for signed overflow only?

That's the goal, yes. Unsigned integers don't overflow by definition; they wrap. The function comment even explains:
```
/// Perform the given integer operation, which is known to need at most BitWidth
/// bits, and check for overflow in the original type (if that type was not an
/// unsigned type).
```

> Basically if the LHS is unsigned don't check for overflow and just return the result of the operation, so this will get diagnosed:
> 
> ```c++
> std::size_t k1Pb1 =  1024 * 1024 * 1024 * 1024 * 1024U;
> ```
> 
> and funny enough gcc seems to follow the same logic: https://godbolt.org/z/bGW1xrEM6

6.3.1.8p1, in part: Otherwise, if the operand that has unsigned integer type has rank greater or equal to the rank of the type of the other operand, then the operand with signed integer type is converted to the type of the operand with unsigned integer type.

So all of these are converted to `unsigned int` and there is no overflow to diagnose.
